cpp/daal/src/algorithms/tsne/tsne_gradient_descent_impl.i

```diff
@@ -0,0 +1,1050 @@
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __INTERNAL_TSNE_GRADIENT_DESCENT_IMPL_I__
+#define __INTERNAL_TSNE_GRADIENT_DESCENT_IMPL_I__
+
+#include "algorithms/tsne/tsne_gradient_descent.h"
+#include "data_management/data/numeric_table.h"
+#include "data_management/data/csr_numeric_table.h"
+#include "services/daal_defines.h"
+#include "services/env_detect.h"
+#include "src/externals/service_math.h"
+#include "src/externals/service_dispatch.h"
+#include "src/services/service_data_utils.h"
+#include "src/services/service_utils.h"
+#include "src/services/service_defines.h"
+#include "src/data_management/service_numeric_table.h"
+#include "src/algorithms/service_error_handling.h"
+#include <iostream>
+#include <ctime>
+#include <chrono>
+#include <fstream>
+
+using namespace daal::data_management;
+using namespace daal::internal;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace internal
+{
+template <typename IdxType, daal::CpuType cpu>
+services::Status maxRowElementsImpl(const size_t * row, const IdxType N, IdxType & nElements)
+{
+    nElements                 = 0;
+    const IdxType nThreads    = threader_get_threads_number();
+    const IdxType nBlocks     = ((nThreads < N) ? nThreads : 1);
+    const IdxType sizeOfBlock = N / nBlocks + !!(N % nBlocks);
+    size_t * maxElements      = services::internal::service_calloc<size_t, cpu>(nThreads);
+    DAAL_CHECK_MALLOC(maxElements);
+
+    daal::threader_for(nBlocks, nBlocks, [&](IdxType iBlock) {
+        const IdxType iStart = iBlock * sizeOfBlock;
+        const IdxType iEnd   = (((iBlock + 1) * sizeOfBlock > N) ? N : iStart + sizeOfBlock);
+        size_t max           = 0;
+        for (IdxType i = iStart; i < iEnd; ++i)
+        {
+            max = services::internal::max<cpu, size_t>(max, row[i + 1] - row[i]);
+        }
+        maxElements[iBlock] = max;
+    });
+
+    for (IdxType i = 0; i < nBlocks; ++i)
+    {
+        nElements = services::internal::max<cpu, IdxType>(nElements, (IdxType)maxElements[i]);
+    }
+
+    services::internal::service_free<size_t, cpu>(maxElements);
+    return services::Status();
+}
+
+template <typename IdxType, typename DataType, daal::CpuType cpu>
+services::Status boundingBoxKernelImpl(DataType * posx, DataType * posy, const IdxType N, const IdxType nNodes, DataType & radius)
+{
+    DAAL_CHECK_MALLOC(posx);
```

Comment:
I don't think these checks are necessary because you've checked them before calling the function